src/Core/Cache/Type/DatabaseCache.php

```diff
@@ -47,17 +47,21 @@ public function getAllKeys(?string $prefix = null): array
 			}
 
 			$stmt = $this->dba->select('cache', ['k'], $where);
+		} catch (\Exception $exception) {
+			throw new CachePersistenceException(sprintf('Cannot fetch all keys with prefix %s', $prefix), $exception);
+		}
 
+		try {
 			$keys = [];
 			while ($key = $this->dba->fetch($stmt)) {
 				array_push($keys, $key['k']);
 			}
 		} catch (\Exception $exception) {
-			throw new CachePersistenceException(sprintf('Cannot fetch all keys with prefix %s', $prefix), $exception);
-		} finally {
```

Comment:
So the existing code was correct, just potentially confusing?

Reply:
No, it was both. PHPStan reports for this code:

`:58    Variable $stmt might not be defined.`

This would be the case if an exceptions is thrown while calling `$stmt = $this->dba->select()`. If this fails $stmt would be undefined. Later, while calling `$key = $this->dba->fetch($stmt)` another exception could be thrown. If thats the case $stmt should be closed before throwing an exeption.

That was fixed by splitting the code into two try blocks: The first block catches exceptions and throw a `CachePersistenceException`. The second block also catches exceptions but will close the $stmt before throwing the `CachePersistenceException`.

If no exception is thrown the $stmt will be closed as well.